File: python/array/152_Maximum_Product_Subarray.py
'''
brute force: Time: O(n^2)
DP
Time: O(n)
Space: O(1)
'''

class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        res = max(nums)  # initial result
        curMax, curMin = 1, 1  # initial current max, current min. Because we have negative numbers

        for n in nums:
            # temp = curMax * n
            # curMax = max(curMax * n, curMin * n, n)
            # curMin = min(temp, curMin * n, n)
            curMax, curMin = max(curMax * n, curMin * n, n),  min(curMax * n, curMin * n, n)
            res = max(res, curMax)

        return res

File: javascript/array/152_Maximum_Product_Subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxProduct = function(nums) {
    let res = nums.reduce((a, b) => Math.max(a, b), -Infinity);
    let curMax = 1, curMin = 1;

    for (let n of nums) {
        temp = curMax * n;
        curMax = Math.max(curMax * n, curMin * n, n)
        curMin = Math.min(temp, curMin * n, n);
        res = Math.max(curMax, res);
    }
    return res;
};